### PR TITLE
RCHAIN-3506: Move creation of deployer auth from reducer to normalizer

### PR DESCRIPTION
**Pull request by tymm**
## Overview
In the previous PR I missed that the deployer auth has to be created in the normalizer.
If it is created in the reducer, the drain vault attack can still be executed.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3506

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by tymm** [comment]
bors try

**Comment by tymm** [comment]
bors try

File: models/src/main/protobuf/RhoTypes.proto [hunk 1]
```diff
@@ -149,7 +153,8 @@ message New {
     Par p = 2 [(scalapb.field).no_box = true];
     // For normalization, uri-referenced variables come at the end, and in lexicographical order.
     repeated string uri = 3;
-    bytes locallyFree = 4 [(scalapb.field).type = "coop.rchain.models.AlwaysEqual[scala.collection.immutable.BitSet]"];
+    DeployerAuth auth = 4;
```
Comment (ArturGajowy) [inline models/src/main/protobuf/RhoTypes.proto]: We should make sure this is only filled when there's a `rho:deployer:auth` in the `uri`-s. Otherwise this field should stay empty (all fields are optional in proto 3)

Comment (ArturGajowy) [reply models/src/main/protobuf/RhoTypes.proto]: I mention the above is just in case :point_up:

Comment (ArturGajowy) [reply models/src/main/protobuf/RhoTypes.proto]: also, let's call it `deployerId`

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -404,40 +400,31 @@ class DebruijnInterpreter[M[_], F[_]](
...
-      )(newEnv: Env[Par], urn: String): Either[ReduceError, Env[Par]] =
-        if (urn == "rho:deployer:auth") {
-          deployParameters.userId.exprs.headOption match {
-            case Some(Expr(GByteArray(bs))) =>
-              newEnv.put(Par(unforgeables = Vector(GDeployerAuth(bs)))).asRight[ReduceError]
-            case _ => ReduceError("No deploy parameters set").asLeft[Env[Par]]
-          }
-        } else
+      def addUrn(newEnv: Env[Par], urn: String): Either[ReduceError, Env[Par]] =
+        if (urn == "rho:deployer:auth")
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: we should change the URI to `rho:rchain:deployerId`. It's how we refer to it all the time after all

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 3]
```diff
@@ -404,40 +400,31 @@ class DebruijnInterpreter[M[_], F[_]](
...
+          neu.auth
+            .map { case DeployerAuth(pk) => newEnv.put(GDeployerAuth(pk)).asRight[ReduceError] }
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: same with DeployerAuth (DeployerId) and GDeployerAuth

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 4]
```diff
@@ -404,40 +400,31 @@ class DebruijnInterpreter[M[_], F[_]](
...
+            .getOrElse(ReduceError(s"No deployer auth set").asLeft[Env[Par]])
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: This should say:
```No DeployId set despite `rho:rchain:deployerId` being used in a term. This is a bug in the normalizer or on the path from it.```
The error should be BugFoundError.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 5]
```diff
@@ -101,7 +100,8 @@ object RemainderNormalizeMatcher {
 
 object CollectionNormalizeMatcher {
   def normalizeMatch[M[_]](c: Collection, input: CollectVisitInputs)(
-      implicit sync: Sync[M]
+      implicit sync: Sync[M],
+      deployerPk: Option[PublicKey]
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: well, shit... I guess let's have the normalizer pass the deployerPk via implicits. Please leave a comment on the normalizer's entry point saying:

```//FIXME before adding any more implicits, or fields to the `*VisitInputs` classes, make the normalizer use ApplicativeAsk / MonadState instead```

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 6]
```diff
@@ -929,24 +927,30 @@ object ProcNormalizeMatcher {
...
-                bodyResult.par.locallyFree.from(newCount).map(x => x - newCount)
-              )
-            ProcVisitOutputs(input.par.prepend(resultNew), bodyResult.knownFree)
+        val newBindings  = sortBindings.map(row => (row._2, row._3, row._4, row._5))
+        val uris         = sortBindings.flatMap(row => row._1)
+        val newEnv       = input.env.newBindings(newBindings.toList)
+        val newCount     = newEnv.count - input.env.count
+        val hasAuth      = uris.contains("rho:deployer:auth")
+        if (hasAuth && deployerPk.isEmpty)
+          NormalizerError("No deployer public key set").raiseError[M, ProcVisitOutputs]
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: This should say:
``` `rho:rchain:deployerId` was used in rholang usage context where DeployerId is not available.```

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 7]
```diff
@@ -929,24 +927,30 @@ object ProcNormalizeMatcher {
...
+        else {
+          val optDeployerAuth =
+            if (hasAuth) deployerPk.map(pk => DeployerAuth(pk.bytes.toByteString))
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: :ok_hand: so my just-in-case comment was not needed :D

I'd though `s/hasAuth/requiresDeployerId`

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/CostAccountingReducerTest.scala [hunk 8]
```diff
@@ -84,13 +84,9 @@ class CostAccountingReducerTest extends FlatSpec with Matchers with TripleEquals
...
     implicit val rand     = Blake2b512Random(128)
     implicit val cost     = CostAccounting.initialCost[Task](Cost(1000)).runSyncUnsafe(1.second)
-    val reducer = new DebruijnInterpreter[Task, Task.Par](
-      tuplespaceAlg,
-      Map.empty,
-      Ref.unsafe[Task, DeployParameters](DeployParameters.empty)
-    )
-    val send = Send(Par(exprs = Seq(GString("x"))), Seq(Par()))
-    val test = reducer.inj(send).attempt.runSyncUnsafe(1.second)
+    val reducer           = new DebruijnInterpreter[Task, Task.Par](tuplespaceAlg, Map.empty)
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/CostAccountingReducerTest.scala]: I think the changes here belong to a different commit.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala [hunk 9]
```diff
@@ -1127,6 +1133,7 @@ class ProcMatcherSpec extends FlatSpec with Matchers {
                 1,
                 Send(EVar(BoundVar(0)), List[Par](GInt(47)), false, BitSet(0)),
                 Vector.empty,
+                None,
                 BitSet()
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: wouldn't you prefer those AST creations to use named parameters? I'd be fine with such a change, but leave it up to you :)

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala [hunk 10]
```diff
@@ -43,6 +44,8 @@ class Rev[A](
     |}
   """.stripMargin
 
+  implicit val deployerPk: Option[PublicKey] = None
```
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala]: What I'd do would be:
- create a new branch, cherry pick the first commit
- do a couple commits replacing the 2nd commit:
  - extract any needed test util methods for calling the interpreter/parbuilder
    - probably make sure to share them across tests
  - pass the deployerId parameter to those methods and elsewhere, do not do any logic with it yet
  - do the logic
- cherry pick the last two commits

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 11]
```diff
@@ -929,24 +927,30 @@ object ProcNormalizeMatcher {
...
-            ProcVisitOutputs(input.par.prepend(resultNew), bodyResult.knownFree)
+        val newBindings  = sortBindings.map(row => (row._2, row._3, row._4, row._5))
+        val uris         = sortBindings.flatMap(row => row._1)
+        val newEnv       = input.env.newBindings(newBindings.toList)
+        val newCount     = newEnv.count - input.env.count
+        val hasAuth      = uris.contains("rho:deployer:auth")
+        if (hasAuth && deployerPk.isEmpty)
+          NormalizerError("No deployer public key set").raiseError[M, ProcVisitOutputs]
+        else {
+          val optDeployerAuth =
```
[the remark this answers is not in the document]
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Done

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala]: Done

File: casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala [hunk 12]
```diff
@@ -28,21 +26,27 @@ object ListenAtName {
 
   private def buildPar[G[_]: BuildPar](name: G[Name]) = implicitly[BuildPar[G]].build(name)
 
-  implicit def buildParF[F[_]: MonadError[?[_], Throwable]] = new BuildPar[λ[A => F[Id[A]]]] {
+  implicit def buildParF[F[_]: MonadError[?[_], Throwable]](
```
[the remark this answers is not in the document]
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala]: Not sure if I understand this comment correctly. Please check if the current state is fine or not. I think it should be fine though.

File: models/src/main/protobuf/RhoTypes.proto (hunk 1 above)
Comment (tymm) [reply models/src/main/protobuf/RhoTypes.proto]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 3 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 4 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala (hunk 5 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala (hunk 6 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala (hunk 7 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Done

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/CostAccountingReducerTest.scala (hunk 8 above)
Comment (tymm) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/CostAccountingReducerTest.scala]: Nope, it is the right one :)

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala (hunk 9 above)
Comment (tymm) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: Agreed and done.

**Comment by tymm** [comment]
bors try

File: rholang/src/main/scala/coop/rchain/rholang/build/CompiledRholangSource.scala [hunk 13]
```diff
@@ -22,7 +22,7 @@ object CompiledRholangSource {
          #$fileContent
          #""".stripMargin('#')
     }
-    override val term: Par = ParBuilder[Coeval].buildNormalizedTerm(code).value()
+    override val term: Par = ParBuilder[Coeval].buildNormalizedTerm(code, None).value()
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/build/CompiledRholangSource.scala]: I've got to say, the results of having the deployerPk a non-default parameter are far better than expected :astonished:. It's *so painfully clear* where we might be relying on it, and yet don't have it. Superb. I also learnt a thing or two about our codebase :joy:

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/ParBuilder.scala [hunk 14]
```diff
@@ -13,30 +14,30 @@ import coop.rchain.rholang.syntax.rholang_mercury.{parser, Yylex}
...
 
 trait ParBuilder[F[_]] {
-  def buildNormalizedTerm(source: String): F[Par]
+  def buildNormalizedTerm(source: String, deployerPk: Option[PublicKey]): F[Par]
 
-  def buildNormalizedTerm(reader: Reader): F[Par]
+  def buildNormalizedTerm(reader: Reader, deployerPk: Option[PublicKey]): F[Par]
 
-  def buildPar(proc: Proc): F[Par]
+  def buildPar(proc: Proc, deployerPk: Option[PublicKey]): F[Par]
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/ParBuilder.scala]: Yeah, passing the deployerPk to the ParBuilder raises eyebrows... I guess next steps in future PRs are:
 - having a RhoEnv data structure passed with the deployerPk and deployId (deploy signature)
 - maybe making RhoEnv a trait, and abstracting away places where non-rholang things leak into rholang.

For now though, this is precisely the result I asked for and I find it much better than the previous version (which also had this issue).

File: casper/src/test/scala/coop/rchain/casper/util/rholang/DeployerAuthTest.scala [hunk 15]
```diff
@@ -67,14 +70,14 @@ class DeployerAuthTest extends FlatSpec with Matchers {
...
-          s"""new auth(`rho:deployer:auth`) in { contract @"checkAuth"(input, ret) = { ret!(*input == *auth) }}""",
+          s"""new auth(`rho:rchain:deployerId`) in { contract @"checkAuth"(input, ret) = { ret!(*input == *auth) }}""",
         timestamp = System.currentTimeMillis(),
         accounting.MAX_VALUE,
         sec = deployer
       )
       val captureChannel = "__RETURN_VALUE__"
       val checkAuth =
-        s"""new auth(`rho:deployer:auth`), ret in {
+        s"""new auth(`rho:rchain:deployerId`), ret in {
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/util/rholang/DeployerAuthTest.scala]: I expected this to happen in `rename deployer auth to deploy id` commit - but no big deal :) I'm bringing this up mainly to make sure you're in control of your rebases :)

Comment (tymm) [reply casper/src/test/scala/coop/rchain/casper/util/rholang/DeployerAuthTest.scala]: Yeah, sorry for missing that. But not a rebasing issue :)

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/ParBuilder.scala (hunk 14 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/ParBuilder.scala]: In the deploy id PR I am introducing `NormailizerEnv`.

But I don't get your second point about abstracting away places where non-rholang things leak into rholang. Can you elaborate?

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/ParBuilder.scala]: We'd like rholang to stay pristine and devoid of any usage-specific trivia, like the deployer id or deploy id. From rholang's perspective, if anything, one can imagine a use case for having compile-time provided values, and maybe runtime-provided values as well. Those are usually provided in Map-s dubbed Environments ;)

IOW, rholang should not know there is a 'deploy id' or a 'deployer id' or even their types. Rather, it should know that there will be some values injected at compile time, and various plugins to rholang (like the one providing support for the deployId name) can resolve those values using the string keys only the plugin need to know. I mean, deployId support is not a plugin now. But should be, if we're to make rholang pristine yet again. Point 2) is for long-term consideration though :)
